lib/maplibre-compose-material3/src/commonMain/kotlin/dev/sargunv/maplibrecompose/material3/controls/ScaleBar.kt

```diff
@@ -1,253 +1,388 @@
   }
+
+private fun metricBarParams(
+  metersPerPx: Float,
+  maxBarLengthPx: Float,
+  m: String,
+  km: String,
+): ScaleBarParams {
+  val stop = findStop(maxBarLengthPx * metersPerPx, METRIC_STOPS)
+  return ScaleBarParams(
+    text =
+      if (stop >= 1000) {
+        "${(stop/1000f).toShortString()} $km"
+      } else {
+        "${stop.toShortString()} $m"
+      },
+    barLengthPx = stop / metersPerPx,
+  )
 }
 
-/**
- * Converts [this] value in meters to the corresponding value in feet
- *
- * @return [this] meters value converted to feet
- */
-private fun Double.toFeet(): Double {
-  return this * CENTIMETERS_IN_METER / CENTIMETERS_IN_INCH / INCHES_IN_FOOT
+private fun yardsBarParams(
+  metersPerPx: Float,
+  maxBarLengthPx: Float,
+  yd: String,
+  mi: String,
+): ScaleBarParams {
+  val stop = findStop((maxBarLengthPx * metersPerPx / METERS_IN_YARD).toFloat(), YARDS_STOPS)
+  return ScaleBarParams(
+    text =
+      if (stop >= YARDS_IN_MILE) {
+        "${(stop/YARDS_IN_MILE).toShortString()} $mi"
+      } else {
+        "${stop.toShortString()} $yd"
+      },
+    barLengthPx = (stop * METERS_IN_YARD / metersPerPx).toFloat(),
+  )
+}
+
+private fun feetBarParams(
+  metersPerPx: Float,
+  maxBarLengthPx: Float,
+  ft: String,
+  mi: String,
+): ScaleBarParams {
+  val stop = findStop((maxBarLengthPx * metersPerPx / METERS_IN_FEET).toFloat(), FEET_STOPS)
+  return ScaleBarParams(
+    text =
+      if (stop >= FEET_IN_MILE) {
+        "${(stop/FEET_IN_MILE).toShortString()} $mi"
+      } else {
+        "${stop.toShortString()} $ft"
+      },
+    barLengthPx = (stop * METERS_IN_FEET / metersPerPx).toFloat(),
+  )
 }
 
 /**
- * Converts [this] value in feet to the corresponding value in miles
- *
- * @return [this] feet value converted to miles
+ * find the largest stop in the list of stops (sorted in ascending order) that is below or equal
+ * [max].
  */
-private fun Double.toMiles(): Double {
-  return this / FEET_IN_MILE
+private fun findStop(max: Float, stops: List<Float>): Float {
+  var maxStop = stops.first()
+  for (stop in stops) {
+    if (stop <= max) maxStop = stop else break
+  }
+  return maxStop
 }
 
-private const val CENTIMETERS_IN_METER: Double = 100.0
-private const val METERS_IN_KILOMETER: Double = 1000.0
-private const val CENTIMETERS_IN_INCH: Double = 2.54
-private const val INCHES_IN_FOOT: Double = 12.0
-private const val FEET_IN_MILE: Double = 5280.0
+/** format like an int if this has no decimals */
+private fun Float.toShortString() = if (this % 1 == 0f) toInt().toString() else toString()
+
+/** list of meters stops */
+private val METRIC_STOPS: List<Float> =
+  listOf(
+    0.1f,
+    0.2f,
+    0.5f,
+    1f,
+    2f,
+    5f,
+    10f,
+    20f,
+    50f,
+    100f,
+    200f,
+    500f,
+    1000f,
+    2000f,
+    5000f,
+    10000f,
+    20000f,
+    50000f,
+    100000f,
+    200000f,
+    500000f,
+    1000000f,
+    2000000f,
+    5000000f,
+    10000000f,
+    20000000f,
+    40000000f,
```

Comment:
these can be generated, something like:

```kt
private fun buildStops(mantissas: Set<Int>, exponents: IntRange): List<Double> = buildList {
    for (e in exponents)
        for (m in mantissas)
            add(m.toDouble() * Math.pow(10.0, e.toDouble()))
}
```

Reply:
Looks quite ingenious! For me at least, this would introduce a readability problem however, as I find it quite hard to understand what would be the resulting list.

This is the sort of complexity where I'd say that this needs a test case. But then, at the end, it will probably not really be less lines of code.

But anyway, if you want to go ahead with this anyway, I don't mind much.

Reply:
> For me at least, this would introduce a readability problem however, as I find it quite hard to understand what would be the resulting list.

I considered that, but I think it's mitigated with just a comment explaining it's based on scientific notation: `m*10^e`.

> if you want to go ahead with this anyway, I don't mind much.

sure, I'll take care of it; just need to tweak the above pattern for the transition between feet/yards and miles.